### PR TITLE
darwin: Retry IOCreatePlugInInterfaceForService on failure

### PR DESCRIPTION
**Pull request by jviotti**
We have a case where a device might not be detected by libusb if we plug
it after initializing the library. We were able to reproduce this issue
pretty consistently on a Mac Mini and several MacBook Pros running very
recent macOS versions (Mojave) using the Balena Fin board
(https://www.balena.io/fin/). I don't know if this happens with other
USB devices.

Enabling debug output revealed the following:

```
[ 7.901582] [00004f07] libusb: debug [darwin_get_cached_device] finding cached device for sessionID 0x2a9b5279f04
[ 7.901764] [00004f07] libusb: debug [darwin_get_cached_device] matching sessionID 0x2a9b5279f04 against cached device with sessionID 0x17dbd029b6a
[ 7.901798] [00004f07] libusb: debug [darwin_get_cached_device] matching sessionID 0x2a9b5279f04 against cached device with sessionID 0x1688763269f
[ 7.901818] [00004f07] libusb: debug [darwin_get_cached_device] matching sessionID 0x2a9b5279f04 against cached device with sessionID 0x4c0aefaa
[ 7.901831] [00004f07] libusb: debug [darwin_get_cached_device] matching sessionID 0x2a9b5279f04 against cached device with sessionID 0x0
[ 7.901845] [00004f07] libusb: debug [darwin_get_cached_device] caching new device with sessionID 0x2a9b5279f04
[ 7.903377] [00004f07] libusb: debug [darwin_device_from_service] could not set up plugin for service: out of resources
[ 8.029152] [00000307] libusb: debug [libusb_get_device_list]
```

The "out of resources" error comes from
`IOCreatePlugInInterfaceForService` which will apparently return this
error if the `Start` method of the interface is not success (see
https://github.com/opensource-apple/IOKitUser/blob/b80a5cbc0ebfb5c4954ef6d757918db0e4dc4b7f/IOCFPlugIn.c#L232).

Retrying `IOCreatePlugInInterfaceForService` makes it work for me.
Signed-off-by: Juan Cruz Viotti <juan@balena.io>

### CODE REVIEW

**Comment by jviotti** [comment]
We don't really know if this is the best solution, and we also don't understand if this is an issue with the Balena Fin board, or macOS, so consider this PR more as a way to start a conversation about how to solve this problem!

**Comment by jviotti** [comment]
The output after the fix is:

```
[ 8.412945] [00004f07] libusb: debug [darwin_get_cached_device] finding cached device for sessionID 0x2d66aa5d76c
[ 8.413152] [00004f07] libusb: debug [darwin_get_cached_device] matching sessionID 0x2d66aa5d76c against cached device with sessionID 0x17dbd029b6a
[ 8.413178] [00004f07] libusb: debug [darwin_get_cached_device] matching sessionID 0x2d66aa5d76c against cached device with sessionID 0x1688763269f
[ 8.413195] [00004f07] libusb: debug [darwin_get_cached_device] matching sessionID 0x2d66aa5d76c against cached device with sessionID 0x4c0aefaa
[ 8.413210] [00004f07] libusb: debug [darwin_get_cached_device] matching sessionID 0x2d66aa5d76c against cached device with sessionID 0x0
[ 8.413224] [00004f07] libusb: debug [darwin_get_cached_device] caching new device with sessionID 0x2d66aa5d76c
[ 8.415190] [00004f07] libusb: debug [darwin_cache_device_descriptor] cached device descriptor:
[ 8.415210] [00004f07] libusb: debug [darwin_cache_device_descriptor]   bDescriptorType:    0x01
[ 8.415221] [00004f07] libusb: debug [darwin_cache_device_descriptor]   bcdUSB:             0x0200
[ 8.415230] [00004f07] libusb: debug [darwin_cache_device_descriptor]   bDeviceClass:       0x00
[ 8.415238] [00004f07] libusb: debug [darwin_cache_device_descriptor]   bDeviceSubClass:    0x00
[ 8.415247] [00004f07] libusb: debug [darwin_cache_device_descriptor]   bDeviceProtocol:    0x00
[ 8.415256] [00004f07] libusb: debug [darwin_cache_device_descriptor]   bMaxPacketSize0:    0x40
[ 8.415265] [00004f07] libusb: debug [darwin_cache_device_descriptor]   idVendor:           0x0a5c
[ 8.415274] [00004f07] libusb: debug [darwin_cache_device_descriptor]   idProduct:          0x2764
[ 8.415283] [00004f07] libusb: debug [darwin_cache_device_descriptor]   bcdDevice:          0x0000
[ 8.415292] [00004f07] libusb: debug [darwin_cache_device_descriptor]   iManufacturer:      0x01
[ 8.417205] [00004f07] libusb: debug [darwin_cache_device_descriptor]   iProduct:           0x02
[ 8.417218] [00004f07] libusb: debug [darwin_cache_device_descriptor]   iSerialNumber:      0x00
[ 8.417224] [00004f07] libusb: debug [darwin_cache_device_descriptor]   bNumConfigurations: 0x01
[ 8.479523] [00004f07] libusb: debug [darwin_check_configuration] active config: 1, first config: 1
[ 8.479566] [00004f07] libusb: debug [process_new_device] allocating new device in context 0x1024183d0 for with session 0x2d66aa5d76c
[ 8.479587] [00004f07] libusb: debug [process_new_device] found device with address 9 port = 1 parent = 0x0 at 0x1023064fa
[ 8.479674] [0000340b] libusb: debug [handle_events] poll() returned 1
[ 8.479698] [0000340b] libusb: debug [handle_events] caught a fish on the hotplug pipe
[ 8.479739] [0000340b] libusb: debug [handle_events] poll() 2 fds with timeout in 0ms
[ 8.479762] [0000340b] libusb: debug [handle_events] poll() returned 0
[ 8.479796] [0000340b] libusb: debug [libusb_get_next_timeout] no URBs, no timeout!
[ 8.479808] [0000340b] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 8.479818] [0000340b] libusb: debug [handle_events] poll() 2 fds with timeout in 60000ms
[10.033212] [00000307] libusb: debug [libusb_get_device_list]
[10.033370] [00000307] libusb: debug [libusb_get_device_descriptor]
```

**Comment by hjelmn** [comment]
I see no reason to not retry. Might be worth sleeping between retries though. Maybe add a nanosleep of 1us. Please also add a usbi_dbg message on each retry.

**Comment by hjelmn** [comment]
Also add a comment on why there is a retry loop. Use C89-style ```/* */``` comments NOT C++/C99 ```//``` comments to maintain consistency in style.

**Comment by jviotti** [comment]
Hey @hjelmn. Thanks for taking a look. I'll make the changes and ping you again :)

**Comment by jviotti** [comment]
@hjelmn Done! What do you think?

File: libusb/os/darwin_usb.c [hunk 1]
```diff
@@ -296,10 +296,24 @@ static usb_device_t **darwin_device_from_service (io_service_t service)
...
+    }
 
-  kresult = IOCreatePlugInInterfaceForService(service, kIOUSBDeviceUserClientTypeID,
-                                              kIOCFPlugInInterfaceID, &plugInInterface,
-                                              &score);
+    usbi_dbg ("set up plugin for service retry: %s", darwin_error_str (kresult));
+
+    /* sleep for a little while before trying again */
+    nanosleep(&(struct timespec){0, 1000}, NULL);
+  }
```
Comment (hjelmn) [inline libusb/os/darwin_usb.c]: I know that struct timespec will probably not change but can you specify the subobjects being initialized here:

```
nanosleep(&(struct timespec){.tv_sec = 0, .tv_nsec = 1000}, NULL);
```

**Comment by hjelmn** [comment]
Address the one thing then I will merge this.

File: libusb/os/darwin_usb.c (hunk 1 above)
Comment (jviotti) [reply libusb/os/darwin_usb.c]: Sorted!

**Comment by jviotti** [comment]
@hjelmn Good point. Done!

**Comment by jviotti** [comment]
@hjelmn One of the macOS Travis CI builds timed out for some reason, and I can't restart it.

**Pull request closed without merging**